engine/filesync/localfs.go

```diff
@@ -260,10 +302,39 @@ func (local *localFS) Sync( //nolint:gocyclo
 
 			default:
 				eg.Go(func() error {
-					appliedChange, err := local.WriteFile(egCtx, kind, path, upperStat, remote)
+					// since we can't really know when a root path is done as
+					// all files are synced in parallel, we track everything
+					// we see and then change the span end time before returning.
+					rootPathSpan := getRootPath(path)
+
+					appliedChange, written, werr := local.WriteFile(ctx, kind, path, upperStat, remote)
+
+					writeEnd := time.Now()
+					m := telemetry.Meter(ctx, "dagger.io/filesync")
+					fsMetric, err := m.Int64Gauge(telemetry.FilesyncWrittenBytes)
 					if err != nil {
 						return err
 					}
+					attrs := []attribute.KeyValue{
+						attribute.String(telemetry.MetricsTraceIDAttr, rootPathSpan.span.SpanContext().TraceID().String()),
```

Comment:
tests showing a panic on this line: https://dagger.cloud/dagger/traces/9f78297444137d035bbc5a52bf06493b?span=10f0f3a825d52517

(just in case you didn't see it already since finding the panic in the engine requires some spelunking around the telemetry)